vault/resource_plugin.go

```diff
@@ -0,0 +1,270 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/vault/api"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/consts"
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+	"github.com/hashicorp/terraform-provider-vault/util"
+)
+
+const (
+	fieldSHA256   = "sha256"
+	fieldCommand  = "command"
+	fieldArgs     = "args"
+	fieldEnv      = "env"
+	fieldOCIImage = "oci_image"
+	fieldRuntime  = "runtime"
+)
+
+var (
+	// Version regex is intentionally loose, its main purpose is to disallow
+	// slashes so they can be used to delineate from the name. Version segment
+	// is optional.
+	pluginIDRegex = regexp.MustCompile(`^(auth|secret|database)(?:/version/([0-9a-zA-Z.-]+?))?/name/(.+)$`)
+)
+
+func pluginFromID(id string) (typ string, name string, version string) {
+	matches := pluginIDRegex.FindStringSubmatch(id)
+	switch {
+	case matches == nil || len(matches) < 3:
+		return "", "", ""
+	case len(matches) == 3:
+		return matches[1], matches[2], ""
+	default:
+		return matches[1], matches[3], matches[2]
+	}
+}
+
+func idFromPlugin(typ, name, version string) string {
```

Comment:
should we add a check here to ensure the user has not passed in a `/` in their version? From the current flow, it looks like we would continue to attempt a plugin with this value. Does Vault return an error if there is a `/` in the version when registering a plugin?

The case I am concerned about is the following:
- a user passes a `/` in the version
- Vault registers the plugin with that value
- TF attempts to read the resource, and fails, because the ID regex does not match

This would cause TF to not be able to set its state, even though there is a resource registered in Vault. The only way to then clean up would be manual deletes in Vault.

Reply:
Vault will error if a provided plugin version can't be parsed as semver, according to the regexes used in [`hashicorp/go-version`](https://github.com/hashicorp/go-version/blob/644291d14038339745c2d883a1a114488e30b702/version.go#L21-L30). IIUC that means there's no need for additional checks here?

Reply:
Yup! Thanks for clarifying